### ISSUE TITLE
Linting should check for Kubernetes Yaml errors

### ISSUE THREAD

**Issue by Grillz** [kubernetes/helm#1741]
Hey I was just deploying a chart and ran into the most annoying error ever:

`json: decNum: got first char '"'`

It turns out that it comes from quoted integers on certain keys, problem is it doesn't give you the first way to find it... and I'm losing hair

`helm lint` should check for these errors otherwise it gives you a false sense of security. Not sure whats so hard about casting a string to an int on Kubernetes end, but helm should provide a way of checking against these **very** tedious errors. Thanks

**Comment by technosophos** [comment]
So the error message in question is coming from the Kubernetes API server. Our plan is to run a Tiller-side validation run in conjunction with the API server (which is required in order to validate types at the server's version). This will be included in Helm 2.2.0.

This morning, we looked into adding this to `helm lint` at your suggestion. But we may not do this because it would require `lint` (an offline operation) to suddenly become an online operation. Still experimenting, though.

**Comment by mariusgrigoriu** [comment]
Whether this is performed by lint or by some other component, this capability would be a big improvement to the chart-developing experience. I think I ran into some cases where invalid yaml would not only fail the deployment, but it would also fail on helm delete. Removing the configmap was the only way to get such charts out of helm list.

**Comment by Grillz** [comment]
@technosophos yeah it looks like a difficult problem to solve the more I dig into it. I think its a worthy goal though, I see two or three of these a day with our current team, and its a pretty bad blocker when you hit it. If you guys need any help on it I know our team would be happy to jump in. Thanks

**Comment by nebril** [comment]
As k8s core repo is already imported to helm, it may be possible to extract validation bits from k8s apiserver, but it can take a lot of effort, as the code in apiserver is quite intertwined and it's not a matter of using one imported function from apiserver.

**Comment by technosophos** [comment]
Server-side validation went live with 2.2.0 yesterday (as part of `helm install|upgrade`). Resources are validated during the prepare phase, which means `--dry-run` will go through validation.

Right now we will not add this to `helm lint` because Lint is an offline operation which doesn't currently need a Kube API server. We can't change that without breaking backward compatibility.

I'll toss this on the 3.0 triage queue so that we can come back to investigate this when we are planning for Helm 3 (which allows breaking changes)

**Comment by zoidbergwill** [comment]
We could use [`kubeval`](https://github.com/garethr/kubeval), but maybe I should try this as a helm plugin first.

**Comment by bacongobbler** [comment]
#2585 should fix this.
